### ISSUE TITLE
Get-PSDrive returns incorrect DisplayRoot property for mounted drive letters in 7.4.1

### ISSUE THREAD

**Issue by sba923** [PowerShell/PowerShell#21064]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When running this for a network-mounted drive letter

```powershell
Get-PSDrive -Name V | Select-Object DisplayRoot
```

The output is incorrect, because the value of the property is incorrect (padded with NULs up to some length):

![image](https://github.com/PowerShell/PowerShell/assets/12860484/0b1dba4f-2411-4c92-935b-32115cce0c12)

The output from Windows PowerShell 5.1 is correct:

![image](https://github.com/PowerShell/PowerShell/assets/12860484/8def126e-5a69-4662-89e9-c723c6a5c59b)

Here's the culprit:

![image](https://github.com/PowerShell/PowerShell/assets/12860484/03b986b8-9d3b-412a-8a99-dd222316aa06)

This is linked to (the fix for) #19903 

### Expected behavior

```console
Get-PSDrive -Name V | Select-Object DisplayRoot

DisplayRoot
-----------
\\pnjnas\V
```

### Actual behavior

```console
Get-PSDrive -Name V | Select-Object DisplayRoot

DisplayRoot
-----------
\\pnjnas\V…
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by jborean93** [comment]
Does the actual property value `(Get-PSDrive -Name V).DisplayRoot` have the `…`. This just seems like how it is formatted by default rather than an incorrect value.

**Comment by sba923** [comment]
> Does the actual property value `(Get-PSDrive -Name V).DisplayRoot` have the `…`. This just seems like how it is formatted by default rather than an incorrect value.

It does contain a bunch of NULs, see the output from `Format-Hex` in the screenshot above.

FWIW, my current workaround in https://gist.github.com/sba923/d0d5ad16f2b12d7785adf830b0395dc2#file-df-ps1 is as follows:

```powershell
$displayroot = $psdrive.DisplayRoot
if ($displayroot.Contains([char]0))
{
    $displayroot = $displayroot.Substring(0,$displayroot.IndexOf([char]0))
}
```

**Comment by jborean93** [comment]
Ah I see sorry, makes more sense now :)

**Comment by jborean93** [comment]
Super weird I can replicate this using the official 7.4.1 release (zip from GitHub release assets) but as soon as I build a local copy of the master branch or the 7.4.1 tag then the problem goes away. I'm not sure what has happened as the PR has been merged and backported and the 7.4.1 tag works just fine when build locally.

**Comment by sba923** [comment]
Interesting. I also can't repro using the local build I use to prepare a PR for #19886...

**Comment by iSazonov** [comment]
This was fixed in #20793 and was backported in #20863. Please re-check v7.4.1 build and report if the issue is still there.

/cc @daxian-dbw It seems there was an unsynchronization with GitHub there.

**Comment by sba923** [comment]
> This was fixed in #20793 and was backported in #20863. Please re-check v7.4.1 build and report if the issue is still there.
> 
> /cc @daxian-dbw It seems there was an unsynchronization with GitHub there.

@iSazonov can you please elaborate? The issue *is* present in the released 7.4.1 MSI...

Is your question addressing the people that were in charge of making it so that the fix would be included in the _released_ 7.4.1?

**Comment by sba923** [comment]
How do I get my working `sba923` fork to be aligned with 7.4.1? 

Building from it currently gets me `PowerShell 7.3.0-preview.3-1215-ge453c006d39a2fa57a903df2d0ea33ddd0cb43dd`

**Comment by iSazonov** [comment]
> Building from it currently gets me `PowerShell 7.3.0-preview.3-1215-ge453c006d39a2fa57a903df2d0ea33ddd0cb43dd`

Try `git fetch --tags --all`.

---

Release process is MSFT internal process. If they skipped something they will fix in 7.4.2.

**Comment by sba923** [comment]
This gets me:

````
PS> git fetch --tags --all
Fetching origin
Fetching upstream
remote: Enumerating objects: 3550, done.
remote: Counting objects: 100% (2293/2293), done.
remote: Compressing objects: 100% (96/96), done.
remote: Total 3550 (delta 2203), reused 2261 (delta 2191), pack-reused 1257
Receiving objects: 100% (3550/3550), 1.73 MiB | 3.37 MiB/s, done.
Resolving deltas: 100% (2671/2671), completed with 615 local objects.
From https://github.com/PowerShell/PowerShell
 * [new branch]          TravisEz13-block-vpack-preview -> upstream/TravisEz13-block-vpack-preview
 * [new branch]          TravisEz13-patch-1     -> upstream/TravisEz13-patch-1
 * [new branch]          TravisEz13-pmc-timeout -> upstream/TravisEz13-pmc-timeout
 * [new branch]          adityapatwardhan-patch-1 -> upstream/adityapatwardhan-patch-1
 * [new branch]          dependabot/nuget/StyleCop.Analyzers-1.2.0-beta.556 -> upstream/dependabot/nuget/StyleCop.Analyzers-1.2.0-beta.556
 * [new branch]          master                 -> upstream/master
 * [new branch]          onboarding-prIssueManagement -> upstream/onboarding-prIssueManagement
 * [new branch]          release/v7.2.18        -> upstream/release/v7.2.18
 * [new branch]          release/v7.3.11        -> upstream/release/v7.3.11
 * [new branch]          release/v7.4.1         -> upstream/release/v7.4.1
 * [new branch]          revert-19152-Webcmdlets-automatic-ContentType-PUT -> upstream/revert-19152-Webcmdlets-automatic-ContentType-PUT
 * [new branch]          source-depot           -> upstream/source-depot
 * [new branch]          spelling-workflow      -> upstream/spelling-workflow
 * [new tag]             v7.0.10                -> v7.0.10
 * [new tag]             v7.0.11                -> v7.0.11
 * [new tag]             v7.0.12                -> v7.0.12
 * [new tag]             v7.0.13                -> v7.0.13
 * [new tag]             v7.1.7                 -> v7.1.7
 * [new tag]             v7.2.10                -> v7.2.10
 * [new tag]             v7.2.11                -> v7.2.11
 * [new tag]             v7.2.12                -> v7.2.12
 * [new tag]             v7.2.13                -> v7.2.13
 * [new tag]             v7.2.14                -> v7.2.14
 * [new tag]             v7.2.15                -> v7.2.15
 * [new tag]             v7.2.16                -> v7.2.16
 * [new tag]             v7.2.17                -> v7.2.17
 * [new tag]             v7.2.18                -> v7.2.18
 * [new tag]             v7.2.3                 -> v7.2.3
 * [new tag]             v7.2.4                 -> v7.2.4
 * [new tag]             v7.2.5                 -> v7.2.5
 * [new tag]             v7.2.6                 -> v7.2.6
 * [new tag]             v7.2.7                 -> v7.2.7
 * [new tag]             v7.2.8                 -> v7.2.8
 * [new tag]             v7.2.9                 -> v7.2.9
 * [new tag]             v7.3.0                 -> v7.3.0
 * [new tag]             v7.3.0-preview.4       -> v7.3.0-preview.4
 * [new tag]             v7.3.0-preview.5       -> v7.3.0-preview.5
 * [new tag]             v7.3.0-preview.6       -> v7.3.0-preview.6
 * [new tag]             v7.3.0-preview.7       -> v7.3.0-preview.7
 * [new tag]             v7.3.0-preview.8       -> v7.3.0-preview.8
 * [new tag]             v7.3.0-rc.1            -> v7.3.0-rc.1
 * [new tag]             v7.3.1                 -> v7.3.1
 * [new tag]             v7.3.10                -> v7.3.10
 * [new tag]             v7.3.11                -> v7.3.11
 * [new tag]             v7.3.2                 -> v7.3.2
 * [new tag]             v7.3.3                 -> v7.3.3
 * [new tag]             v7.3.4                 -> v7.3.4
 * [new tag]             v7.3.5                 -> v7.3.5
 * [new tag]             v7.3.6                 -> v7.3.6
 * [new tag]             v7.3.7                 -> v7.3.7
 * [new tag]             v7.3.8                 -> v7.3.8
 * [new tag]             v7.3.9                 -> v7.3.9
 * [new tag]             v7.4.0                 -> v7.4.0
 * [new tag]             v7.4.0-preview.1       -> v7.4.0-preview.1
 * [new tag]             v7.4.0-preview.2       -> v7.4.0-preview.2
 * [new tag]             v7.4.0-preview.3       -> v7.4.0-preview.3
 * [new tag]             v7.4.0-preview.4       -> v7.4.0-preview.4
 * [new tag]             v7.4.0-preview.5       -> v7.4.0-preview.5
 * [new tag]             v7.4.0-preview.6       -> v7.4.0-preview.6
 * [new tag]             v7.4.0-rc.1            -> v7.4.0-rc.1
 * [new tag]             v7.4.1                 -> v7.4.1
````

The result of the build is `PowerShell 7.4.0-rc.1-182-g6df3809104f009ce943e0a77440f14e9673e3239`, not a 7.4.1 something.

And this issue doesn't repro with the output of that build.

**Comment by jborean93** [comment]
Thanks to @seeminglyscience we've found the problem. The changes in the PR mentioned are definitely in 7.4.1, you can see them when decompiling the dll in the release

![image](https://github.com/PowerShell/PowerShell/assets/8462645/e8cf9a35-3276-4287-a794-60aa208e0d1b)

It uses the `Span<char>` that was used in https://github.com/PowerShell/PowerShell/pull/20793. When stepping through this code at runtime I noticed that `ref bufferSize` (`ref num` in decompiled code) never changed and was always the size of the span/buffer being passed in.

The docs for [WNetGetConnection](https://learn.microsoft.com/en-us/windows/win32/api/winnetwk/nf-winnetwk-wnetgetconnectionw) state for this argument

> Pointer to a variable that specifies the size of the buffer pointed to by the lpRemoteName parameter, in characters. If the function fails because the buffer is not large enough, this parameter returns the required buffer size.

So this value will only be set to the length if the initial buffer wasn't large enough. With a debug build you get when building locally the code has https://github.com/PowerShell/PowerShell/blob/e453c006d39a2fa57a903df2d0ea33ddd0cb43dd/src/System.Management.Automation/engine/Interop/Windows/WNetGetConnection.cs#L29-L34 which set the buffer to 3 explicitly and not the span size so it will actually hit this logic. In a release build it will not. The fix for this is to slice the span to the first null char like in the workaround and really remove the `#if DEBUG` logic so we are actually testing the same code that is run in a release build.

**Comment by jborean93** [comment]
Opened https://github.com/PowerShell/PowerShell/pull/21068

**Comment by sba923** [comment]
> Thanks to @SeeminglyScience we've found the problem. [...] The fix for this is to slice the span to the first null char like in the workaround and really remove the `#if DEBUG` logic so we are actually testing the same code that is run in a release build.

Nice! Will have to live with my workaround until 7.4.2 😜

**Comment by sba923** [comment]
FYI problem is also present in 7.5.0-preview.1

**Comment by sba923** [comment]
FYI problem is also present in 7.5.0-preview.2

**Comment by sba923** [comment]
FYI problem is also present in 7.5.0-preview.3

**Comment by sba923** [comment]
FYI problem is also present in 7.5.0-preview.4

**Comment by sba923** [comment]
FYI problem is also present in 7.5.0-preview.5.

How come @jborean93  given that https://github.com/PowerShell/PowerShell/pull/21068 is closed?

**Comment by sba923** [comment]
What can *I* do about this?

**Comment by sba923** [comment]
As discussed earlier, the issue is still present in 7.5.0-rc.1.

What can I do to help getting this fixed?

**Comment by jborean93** [comment]
I've reopened a new PR based on the original one https://github.com/PowerShell/PowerShell/pull/24600. Hopefully the pwsh team can review it and merge it in.

**Comment by sba923** [comment]
I confirm this is fixed in 7.6.0-preview.2.

**Comment by msftrncs** [comment]
Was this going to be included in releases such as 7.5.3 or 7.5.4?

**Comment by jborean93** [comment]
See https://github.com/PowerShell/PowerShell/pull/25311#issuecomment-2881625618, the backport was opened but was rejected without a customer impact statement so presently it will only be present on 7.6+. cc @TravisEz13
